### ISSUE TITLE
Bug: no-restricted-imports doesn't follow exclusion rules

### ISSUE THREAD

**Issue by mycarrysun** [eslint/eslint#19466]
### Environment

Environment
Node version: 22
npm version: 10.8.2
Local ESLint version: 9.16.0
Global ESLint version:
Operating System: mac os

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```js
rules: {
        "no-restricted-imports": [
          "error",
          {
            patterns: [
              {
                group: [
                  "@org/*/**",
                  "!@org/package-one/allowed/path",
                ],
                message:
                  "Only import from the top level of @org packages",
              },
            ],
          },
        ],
      },
```
</details>

```js
import { someModule } from "@org/package-one/allowed/path";
```

### What did you expect to happen?

I would expect you to be able to import from the excluded patterns.

### What actually happened?

There is an eslint error when using an excluded pattern:
```
'@org/package-one/allowed/path' import is restricted from being used by a pattern. Only import from the top level of @org packages.  ([no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports))
```

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1pbXBvcnRzOiBbXCJlcnJvclwiLCB7IHBhdHRlcm5zOiBbe1xuICAgIGdyb3VwOiBbXCJAb3JnLypcXC8qKlwiLCBcIiFAb3JnL3BhY2thZ2Utb25lL2FsbG93ZWQvcGF0aFwiXSxcbiAgICBtZXNzYWdlOiBcIk9ubHkgaW1wb3J0IGZyb20gdGhlIHRvcCBsZXZlbCBvZiBAb3JnIHBhY2thZ2VzLlwiXG59XX1dKi9cblxuaW1wb3J0IHsgc29tZU1vZHVsZSB9IGZyb20gXCJAb3JnL3BhY2thZ2Utb25lL2FsbG93ZWQvcGF0aFwiOyIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==

### Participation

- [x] I am willing to submit a pull request for this issue.

**Comment by mycarrysun** [comment]
For example we have a built typescript directory called `lib` and I want to disallow importing directly from that directory or any subdirectories, unless its in the exclusion list.
Adding `@org/package-name/lib` to the exclusion list would then allow imports from the path which is undesirable. 

What is the magic sauce to get this to work the way I'm wanting?

**Comment by mdjermanovic** [comment]
> For example we have a built typescript directory called `lib` and I want to disallow importing directly from that directory or any subdirectories, unless its in the exclusion list. Adding `@org/package-name/lib` to the exclusion list would then allow imports from the path which is undesirable.
> 
> What is the magic sauce to get this to work the way I'm wanting?

Can you provide a few examples of what imports should and what imports shouldn't be allowed?

**Comment by mycarrysun** [comment]
> Can you provide a few examples of what imports should and what imports shouldn't be allowed?

I updated the playground example with a few more examples of what might be allowed and what is not allowed: https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1pbXBvcnRzOiBbXCJlcnJvclwiLCB7IHBhdHRlcm5zOiBbe1xuICAgIGdyb3VwOiBbXCJAb3JnLypcXC8qKlwiLCBcIiFAb3JnL3BhY2thZ2Utb25lL2xpYi9hbGxvd2VkL3BhdGhcIiwgXCIhQG9yZy9wYWNrYWdlLXR3by9saWIvYWxsb3dlZC9wYXRoL3R3b1wiXSxcbiAgICBtZXNzYWdlOiBcIk9ubHkgaW1wb3J0IGZyb20gdGhlIHRvcCBsZXZlbCBvZiBAb3JnIHBhY2thZ2VzLlwiXG59XX1dKi9cblxuaW1wb3J0IHsgYWxsb3dlZEV4cG9ydCB9IGZyb20gJ0BvcmcvcGFja2FnZS1vbmUnO1xuaW1wb3J0IHsgYWxsb3dlZEV4cG9ydFR3byB9IGZyb20gJ0BvcmcvcGFja2FnZS10d28nO1xuaW1wb3J0IHsgYWxsb3dlZEV4cG9ydFRocmVlIH0gZnJvbSBcIkBvcmcvcGFja2FnZS1vbmUvbGliL2FsbG93ZWQvcGF0aFwiO1xuaW1wb3J0IHsgZGlzYWxsb3dlZCB9IGZyb20gJ0BvcmcvcGFja2FnZS1vbmUvbGliJztcbmltcG9ydCB7IGRpc2FsbG93ZWRUd28gfSBmcm9tICdAb3JnL3BhY2thZ2UtdHdvL2xpYi9hbGxvd2VkL3BhdGgvdHdvJzsiLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

**Comment by mdjermanovic** [comment]
> > Can you provide a few examples of what imports should and what imports shouldn't be allowed?
> 
> I updated the playground example with a few more examples of what might be allowed and what is not allowed: https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1pbXBvcnRzOiBbXCJlcnJvclwiLCB7IHBhdHRlcm5zOiBbe1xuICAgIGdyb3VwOiBbXCJAb3JnLypcXC8qKlwiLCBcIiFAb3JnL3BhY2thZ2Utb25lL2xpYi9hbGxvd2VkL3BhdGhcIiwgXCIhQG9yZy9wYWNrYWdlLXR3by9saWIvYWxsb3dlZC9wYXRoL3R3b1wiXSxcbiAgICBtZXNzYWdlOiBcIk9ubHkgaW1wb3J0IGZyb20gdGhlIHRvcCBsZXZlbCBvZiBAb3JnIHBhY2thZ2VzLlwiXG59XX1dKi9cblxuaW1wb3J0IHsgYWxsb3dlZEV4cG9ydCB9IGZyb20gJ0BvcmcvcGFja2FnZS1vbmUnO1xuaW1wb3J0IHsgYWxsb3dlZEV4cG9ydFR3byB9IGZyb20gJ0BvcmcvcGFja2FnZS10d28nO1xuaW1wb3J0IHsgYWxsb3dlZEV4cG9ydFRocmVlIH0gZnJvbSBcIkBvcmcvcGFja2FnZS1vbmUvbGliL2FsbG93ZWQvcGF0aFwiO1xuaW1wb3J0IHsgZGlzYWxsb3dlZCB9IGZyb20gJ0BvcmcvcGFja2FnZS1vbmUvbGliJztcbmltcG9ydCB7IGRpc2FsbG93ZWRUd28gfSBmcm9tICdAb3JnL3BhY2thZ2Utb25lL2xpYi9hbGxvd2VkJztcbmltcG9ydCB7IGFsbG93ZWRFeHBvcnRGb3VyIH0gZnJvbSAnQG9yZy9wYWNrYWdlLXR3by9saWIvYWxsb3dlZC9wYXRoL3R3byc7Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19

Here's an example of how it can be configured:

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1pbXBvcnRzOiBbXCJlcnJvclwiLCB7IHBhdHRlcm5zOiBbe1xuICAgIGdyb3VwOiBbXG4gICAgICBcIkBvcmcvKlxcLyoqXCIsXG4gICAgICBcIiFAb3JnL3BhY2thZ2Utb25lL2xpYi9cIixcbiAgICAgIFwiIUBvcmcvcGFja2FnZS1vbmUvbGliL2FsbG93ZWQvXCIsXG4gICAgICBcIiFAb3JnL3BhY2thZ2Utb25lL2xpYi9hbGxvd2VkLyoqXCIsXG4gICAgICBcIiFAb3JnL3BhY2thZ2UtdHdvL2xpYi9cIixcbiAgICAgIFwiIUBvcmcvcGFja2FnZS10d28vbGliL2FsbG93ZWQvXCIsXG4gICAgICBcIiFAb3JnL3BhY2thZ2UtdHdvL2xpYi9hbGxvd2VkLyoqXCIsIFxuICAgIF0sXG4gICAgbWVzc2FnZTogXCJPbmx5IGltcG9ydCBmcm9tIHRoZSB0b3AgbGV2ZWwgb2YgQG9yZyBwYWNrYWdlcy5cIlxufV19XSovXG5cbmltcG9ydCB7IGFsbG93ZWRFeHBvcnQgfSBmcm9tICdAb3JnL3BhY2thZ2Utb25lJztcbmltcG9ydCB7IGFsbG93ZWRFeHBvcnRUd28gfSBmcm9tICdAb3JnL3BhY2thZ2UtdHdvJztcbmltcG9ydCB7IGFsbG93ZWRFeHBvcnRUaHJlZSB9IGZyb20gXCJAb3JnL3BhY2thZ2Utb25lL2xpYi9hbGxvd2VkL3BhdGhcIjtcbmltcG9ydCB7IGRpc2FsbG93ZWQgfSBmcm9tICdAb3JnL3BhY2thZ2Utb25lL2xpYic7XG5pbXBvcnQgeyBkaXNhbGxvd2VkVHdvIH0gZnJvbSAnQG9yZy9wYWNrYWdlLW9uZS9saWIvYWxsb3dlZCc7XG5pbXBvcnQgeyBhbGxvd2VkRXhwb3J0Rm91ciB9IGZyb20gJ0BvcmcvcGFja2FnZS10d28vbGliL2FsbG93ZWQvcGF0aC90d28nOyIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==

**Comment by mycarrysun** [comment]
@mdjermanovic interesting, thanks for finding me a working solution! Didn't think about using a trailing slash to get that to work.
